src/main/java/io/divolte/server/kafka/KafkaFlusher.java

```diff
@@ -81,8 +86,9 @@ public ProcessingDirective process(final Queue<AvroRecordBuffer> batch) {
             logger.debug("Processing batch of {} records.", batchSize);
             final List<ProducerRecord<DivolteIdentifier, AvroRecordBuffer>> kafkaMessages =
                     batch.stream()
+                         .map(i -> i.payload)
                          .map(this::buildRecord)
-                         .collect(Collectors.toList());
+                         .collect(Collectors.toCollection(() -> new ArrayList<>(batchSize)));
```

Comment:
I looked into trying to reduce the number of collections we allocate and just pass streams around. However this wasn't possible:
 - We need to convert each event we're going to send into a Kafka message.
 - Each message we send to Kafka has its own `Future` to report the result.
 - On failure, we need to go back and correlate failures with the Kafka message, so we can preserve them for retry.

(It would be handy if we had (`.zip(…)`) to help with this. There might be a better way, but old-fashioned looping with an index works.)

So the best I could come up with for now was a list for:
 - The Kafka messages we're going to send.
 - The future responses for the messages in flight.
 - The failed messages, which need to be retried.

(The first two are pre-sized, the last isn't but Guava optimises for the common case.)

An alternative I considered was collecting the messages and corresponding futures into a single collection of tuples, albeit a custom container because we don't do tuples. However this is even more allocation-heavy as far as I can tell.

Something I didn't consider but would be slightly lighter-weight is native arrays instead of collections.